src/codec.rs

```diff
@@ -553,14 +572,37 @@ impl<T: Decode> Decode for Vec<T> {
 	fn decode<I: Input>(input: &mut I) -> Result<Self, Error> {
 		<Compact<u32>>::decode(input).and_then(move |Compact(len)| {
 			let len = len as usize;
-			if let IsU8::Yes = <T as Decode>::IS_U8 {
-				let mut r = vec![0; len];
+			let max_preallocation = input.remaining_len()?;
 
-				input.read(&mut r[..len])?;
-				let r = unsafe { core::mem::transmute::<Vec<u8>, Vec<T>>(r) };
+			if let IsU8::Yes = <T as Decode>::IS_U8 {
+				let r = if len <= max_preallocation {
```

Comment:
If we have an `u8` and the rest length does not match the length we have decoded, the input data is corrupt or someone is trying to fool us.

Reply:
oh yes